thunder/core/rematerialization.py

```diff
@@ -538,18 +538,15 @@ def match_fw_and_bw_saved_for_bw_proxies(
         new_required_for_bakward_fw_to_bw_map: Dict[str, Proxy]: mapping of bw names to forward proxies
     """
 
-    old_saved_for_backward_fw = (*fw_trace.bound_symbols[-1].args[1][0], *fw_trace.bound_symbols[-1].args[1][1])
-    old_saved_for_backward_bw = []
+    old_saved_for_backward_fw = [*fw_trace.output[1][0], *fw_trace.output[1][1]]
+    old_saved_for_backward_bw = [None] * len(old_saved_for_backward_fw)
```

Comment:
This line can be simple attribute access on TraceCtx.args instead of inspecting inputs of unpack sequence:
```suggestion
    old_saved_for_backward_bw = (*bw_trace.args[0][0], *bw_trace.args[0][1])
```
If proxies of bw_trace.args do not match the proxies in the unpack sequence operation it means there's a bug somewhere in the preceding transformations of the trace. They should be kept in sync.

Reply:
Hey @IvanYashchuk, thanks for the pointers! 
```
If proxies of bw_trace.args do not match the proxies in the unpack sequence operation it means there's a bug somewhere in the preceding transformations
```
is exactly the issue here, the `TraceSubstitutionProcessor ` will update the names of proxies in the trace, and since the forward and backward are updated separately, they fall out of sync.

It seems the main issue is that the unpacks are a boundary for updating names, since it can't connect the unpacked proxy to its corresponding trace arg. Any ideas how this should be handled properly? Im betting there's something missing from the `TraceSubstitutionProcessor ` but not sure.

Reply:
Please do not rely on names for matching. They are not guaranteed to be the same across retracing.
There is a clear expectation that the saved for backwards returned by the forward are compatible with the ones consumed by the backward, but expecting name stability is asking for trouble.